src/Compilers/Core/Portable/NativePdbWriter/ComMemoryStream.cs

```diff
@@ -52,6 +53,40 @@ public void CopyTo(Stream stream)
             }
         }
 
+        public IEnumerable<ArraySegment<byte>> GetChunks()
+        {
+            byte[] lazyZeroChunk = null;
+
+            int chunkIndex = 0;
+            int remainingBytes = _length;
+            while (remainingBytes > 0)
+            {
+                int bytesToCopy = Math.Min(ChunkSize, remainingBytes);
+
+                byte[] chunk;
+                if (chunkIndex < _chunks.Count)
+                {
+                    chunk = _chunks[chunkIndex++];
+                }
+                else if (bytesToCopy == ChunkSize)
+                {
+                    chunk = lazyZeroChunk ?? (lazyZeroChunk = new byte[ChunkSize]);
```

Comment:
Is this safe? The returned value is mutable hence possible this caching can be observed.

Reply:
It's only used internally. But you are right, it would be better to be safe. Now that I'm thinking about it again we could just allocate an array of size remainingBytes. That could potentially be big, on the other hand in practice this should never happen. It would mean that the PDB has a lot of trailing NULs, which it doesn't. 

I'll simplify the code and remove the caching.

---
In reply to: [136211779](https://github.com/dotnet/roslyn/pull/21712#discussion_r136211779) [](ancestors = 136211779)